rewrite workflow sharing actions into API, add Vue, add social media sharing

anuprulez opened the pull request:
This pull request rewrites the "share workflow" mako in VueJS and enables a workflow to be shareable on social media. For now, we include Twitter, Facebook, Google+ and LinkedIn where the link of any workflow can be posted through social media share option.

![ssw1](https://user-images.githubusercontent.com/3022518/36112324-98e50964-1029-11e8-9592-baf3fee24310.png)

![ssw2](https://user-images.githubusercontent.com/3022518/36112329-9e2b4406-1029-11e8-8000-7006f6e50d42.png)

To share a workflow on multiple social media sites it uses a vue component defined in the [library](https://github.com/nicolasbeauvais/vue-social-sharing). 

Please leave us your comments and feedback!

Thanks @bgruening

anuprulez committed:
Initial commit for share workflows

anuprulez committed:
Add share workflow feature

anuprulez committed:
Add workflow export feature

anuprulez committed:
Add sections for exporting to myexperiment

anuprulez committed:
Add section for add public username

anuprulez committed:
Add share workflow on Twitter feature

anuprulez committed:
Merge branch 'social_share_workflows' of https://github.com/bgruening/galaxy into social_share_workflows

anuprulez committed:
Add anchor link

anuprulez committed:
Add more social share icons and move action buttons to top right

anuprulez committed:
Update the social links

anuprulez committed:
Update the remove wf link and wf sharing mako file

anuprulez committed:
Update return url

anuprulez committed:
Merge branch 'dev' of https://github.com/bgruening/galaxy into social_share_workflows

anuprulez committed:
Get vue-social-sharing through yarn

anuprulez committed:
Merge with dev

anuprulez committed:
add updated yarn file

anuprulez commented:
There are 3 tests failing for API test namely: api.test_workflows.WorkflowsApiTestCase.test_run_rename_based_on_input_collection
api.test_workflows.WorkflowsApiTestCase.test_run_rename_based_on_input_conditional 
api.test_workflows.WorkflowsApiTestCase.test_workflow_metadata_validation_0

Summary of API tests:

![api_test_results](https://user-images.githubusercontent.com/3022518/36162653-482f0042-10e7-11e8-827d-42fa8e7aa42d.png)

But, when I run the API test locally using `./run_tests.sh -api`, all the above tests which fail at the build pass including all other API tests.

![metadata](https://user-images.githubusercontent.com/3022518/36162796-bd61180a-10e7-11e8-9495-9f925020c8e0.png)
![test_run](https://user-images.githubusercontent.com/3022518/36162797-bd8180cc-10e7-11e8-8554-3e1d5c267a12.png)

![screenshot from 2018-02-13 18-01-21](https://user-images.githubusercontent.com/3022518/36162900-01a3b914-10e8-11e8-8907-0a5ea942fff1.png)

I am not sure about the reason why this could happen?

martenson commented:
I re-started the tests, they should pass now. It was not related to this PR.

anuprulez committed:
Show social share icons only when a workflow is importable

anuprulez committed:
Merge branch 'dev' of https://github.com/bgruening/galaxy into social_share_workflows

anuprulez commented:
I added a clause to show the social media share links only when a workflow is importable i.e. when it becomes accessible via a link. Thanks @bgruening for this feedback :)

@martenson does this PR look ok to you or do you have any comments? Thank you :)

martenson commented:
@anuprulez I like the functionality and I will be happy to review, but am now busy with the release-related work. Afterwards I will offer feedback.

bgruening committed:
Merge branch 'dev' into social_share_workflows

guerler commented on client/galaxy/scripts/mvc/workflow/workflow.js:
```diff
@@ -134,7 +134,7 @@ const WorkflowItemView = Backbone.View.extend({
             }">Edit</a></li><li><a href="${Galaxy.root}workflows/run?id=${this.model.id}">Run</a></li><li><a href="${
                 Galaxy.root
-            }workflow/sharing?id=${this.model.id}">Share</a></li><li><a href="${Galaxy.root}api/workflows/${
+            }workflows/share_workflow?id=${this.model.id}">Share</a></li><li><a href="${Galaxy.root}api/workflows/${
```
Should probably be `workflows/share` for consistency.

guerler commented on lib/galaxy/webapps/galaxy/controllers/workflow.py:
```diff
@@ -319,6 +319,13 @@ def share(self, trans, id, email="", use_panels=False):
+    @web.expose
+    @web.require_login("Share or export Galaxy workflows")
+    @web.json
+    def sharing_workflow(self, trans, id, **kwargs):
```
I think, we should use an api endpoint and remove this controller endpoint if possible.

guerler commented on lib/galaxy/webapps/galaxy/controllers/workflow.py:
```diff
@@ -472,7 +491,8 @@ def gen_image(self, trans, id):
             message = 'Galaxy is unable to create the SVG image. Please check your workflow, there might be missing tools.'
-            return trans.fill_template("/workflow/sharing.mako", use_panels=True, item=stored, status=status, message=message)
+            url = url_for('/') + 'workflows/share_workflow?id=' + id + '&status=' + status + '&message=' + message
+            return trans.response.send_redirect(url)
```
If we use an api endpoint, redirecting should happen through the client.

anuprulez committed:
Fix PR review comments

anuprulez committed:
Merge with main

anuprulez committed:
Fix import error

anuprulez committed:
Fix import order

guerler commented on client/galaxy/scripts/mvc/workflow/workflow.js:
```diff
@@ -150,7 +150,7 @@ const WorkflowItemView = Backbone.View.extend({
             }">Run</a></li><li><a id="copy-workflow" style="cursor: pointer;">Copy</a></li><li><a class="link-confirm-shared-${
                 this.model.id
-            }" href="${Galaxy.root}workflow/sharing?unshare_me=True&id=${this.model.id}">Remove</a></li></ul>`;
+            }" href="${Galaxy.root}api/workflows/share_workflow?unshare_me=True&id=${this.model.id}">Remove</a></li></ul>`;
```
Endpoint should probably be `api/workflows/sharing` and re-use https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/web/base/controller.py#L1400 if possible.

guerler commented on lib/galaxy/webapps/galaxy/buildapp.py:
```diff
@@ -138,6 +138,7 @@ def app_factory(global_conf, load_app_kwds={}, **kwargs):
     webapp.add_client_route('/workflows/create')
     webapp.add_client_route('/workflows/run')
     webapp.add_client_route('/workflows/import_workflow')
+    webapp.add_client_route('/workflows/share')
```
Recommend this to be `workflows/sharing` for consistency. This has just recently changed.

anuprulez committed:
Fix review comments and merge with dev

anuprulez commented:
I am not sure whether the failing API tests are related

martenson commented:
I think they are, in the next 10 days there is GCC going on so responses might be delayed.

anuprulez committed:
Add api url

jxtx commented on client/galaxy/scripts/components/ShareWorkflow.vue:
```diff
@@ -0,0 +1,357 @@
+                <social-sharing :url="workflowUrl"
+                                title="Link to a Galaxy workflow: "
+                                quote="Galaxy workflow"
+                                hashtags="usegalaxy"
```
Please don't make this "hashtag" a default as it crufts up the feed.

anuprulez committed:
Remove usegalaxy hashtag

bgruening commented:
We could use `#sciworkflows`: https://twitter.com/smllmp/status/1019981835877736448

anuprulez committed:
Update hashtag

martenson committed:
remove superfluous decorator and corresponding client parsing

martenson committed:
re-expose erroneously hidden endpoint

mvdbeek commented on lib/galaxy/webapps/galaxy/api/workflows.py:
```diff
@@ -437,6 +438,69 @@ def delete(self, trans, id, **kwd):
+    @expose_api
+    def sharing(self, trans, id, **kwargs):
+        """
+        GET /api/workflows/sharing
```
Clearly this method is doing more than `Return a workflow with the given id.`, so I'd like to see those things properly documented
I'm not a fan of copy-pasting controller stuff, which make very few guarantees about being stable, into the API, which should be there permanently across releases. To me all those predicates in kwargs should ideally be separate routes, and for instance returning the workflow list when requesting `unshare_me` is suboptimal.

martenson replied:
I have just started reviewing this, more changes will come. Thanks for the ideas!

martenson committed:
improve documentation

martenson committed:
implement workflow unshare_me as a separate API endpoint

martenson committed:
remove unnecessary encapsulation, fix indentation

martenson committed:
indent html template properly and auto-format js

martenson committed:
use manager to retrieve workflow

move the legacy consistency check into manager also

martenson committed:
pull out the to_dict method to the workflow manager

martenson committed:
use 'scoped' keyword to limit css to the component

make function declaration more explicit

martenson committed:
remove reference to a deleted template

martenson committed:
redesign and refactor the frontend template

martenson committed:
redesign the ui, refactor the svg download

martenson commented:
I am moving this to 19.01, there is still work left to do with the API refactoring that I am working on but it shouldn't hold the release.

jmchilton commented:
Marking as WIP this and delaying to 19.05, if it gets into 19.01 though that would be fantastic.

anuprulez committed:
Fix conflicts and merge with dev

anuprulez committed:
Merge branch 'dev' of github.com:bgruening/galaxy into social_share_workflows

anuprulez committed:
Remove usused whiteline

anuprulez committed:
Remove usused whiteline

anuprulez committed:
Fix py lint errors

anuprulez committed:
Fix failing selenium test

anuprulez commented:
API test is failing for one test which passing in my local

![test_nested_workflow](https://user-images.githubusercontent.com/3022518/49946712-d640a600-feef-11e8-8aae-be3d1039dcf2.png)

![sdfsdf](https://user-images.githubusercontent.com/3022518/49946774-f7a19200-feef-11e8-8fa0-5544055c5b38.png)

Update: Now this test has passed :)

anuprulez commented:
@martenson I rebased this branch with galaxy dev and fixed a bug related to failing selenium test. Can you say what API refactoring you wish to do for workflows? I have some time and I can work on them so that we can ship this feature with 19.01 release.

thank you for all the help.

Update: all tests passed :)

nsoranzo commented:
> Update: all tests passed :)

Yeah, timeout errors in API tests are usually random and go away when re-running.

martenson committed:
Merge branch 'social_share_workflows' of github.com:bgruening/galaxy into social_share_workflows

martenson committed:
fix vue template

martenson committed:
fix styling of bootstrap buttons

martenson commented:
We discussed this privately with @anuprulez. 

I think for the social sharing part we should 
* approach this differently and reuse what @guerler added for other Galaxy objects (the Sharing.vue component from https://github.com/galaxyproject/galaxy/pull/6048)
* but make it more modular 'micro' components (like RuleComponent https://github.com/galaxyproject/galaxy/blob/dev/client/galaxy/scripts/components/RuleCollectionBuilder.vue#L750)

But ultimately also refactor this 'shared sharing' API endpoints to be more atomic for all actions across all objects.

Therefore we are closing this and will start work on the next iteration.

cc @jmchilton @Nerdinacan
